runbooks/CI_pipelines/adding-ci-dockerhub-users.md

```diff
@@ -0,0 +1,84 @@
+
+- see [Using Docker Authenticated Pulls](https://circleci.com/docs/2.0/private-images/)
+
+Fortunately, we have access to a DockerHub organisation in the Ministry of Justice, which has free unlimited pulls of images.
+
+## Security
+
+We are not using [CircleCI contexts](https://circleci.com/docs/2.0/contexts/) as this is a shared organisational level construct, that any pipeline could use.
+We're not set up with context restrictions or groups in CircleCI at time of writing.
+
+We need to ensure that each of the circleci projects has an individual DockerID and token added in [project level environment variables](https://circleci.com/docs/2.0/env-vars/#setting-an-environment-variable-in-a-project), reducing blast radius should it become compromised.
+
+## Prerequisites
+
+You will need:
+
+- Your own Docker ID, which has been:
+  - Added to the `ministryofjustice` dockerhub organisation as an owner.
+  - added to the `opgdockerhubusers` team.
+- Access to the OPG webops team google group email.
+- Access to the project settings on the CircleCI project you are working on.
+
+Ask one of your webops colleagues to help if you don't have details of these.
+
+## Steps
+
+there are 3 main steps:
+
+1. Create a service Docker ID
+2. Add service user to docker organisation and team
+3. Add docker hub credentials to CircleCI
+
+### 1. Create a service Docker ID
+
+1. On hub.docker.com create a Docker ID, using:
+   - A strong randomised password
+   - A name that makes sense for the pipeline.
+   - The email associated with OPG webops group, using a an `+` email alias to associate with the new Docker ID.
+2. Take note of the above credentials and store in the webops secure vault for safekeeping.
+3. Log out of hub.docker.com.
+4. This will send a validation email to the opg team google group, but look out for one with the correct alias.  
+5. Validate the email.
+6. Once validated successfully, log back in with the new credentials
+7. Go to the profile name, drop down to `Account Settings` and select `Security` on the screen.
+8. In the Access Tokens click `New Access Token`.
+9. Give the token a useful name e.g. `CircleCI access token` and click `Create`
+10. on the next Dialog, Copy the access token provided.
+11. paste into the relevant secure vault entry for safekeeping.
+12. Click `Copy and Close`.
+
+### 2. Add the service account to organisation and team
+
+1. Log in to your own DockerID on hub.docker.com.
+2. under Organisations click `ministryofjustice` and then click `Add Member`
+3. Enter the DockerID of the new service user and select `opgdockerhubusers` from the dropdown
+4. Click `Add`.
+
+This user will have member permissions, which will be enough to pull images.
+
+### 3. Add dockerhub credentials to CircleCI
+
+1. go to the project settings of the circleci build pipeline
+2. go to environment variables
+3. Add an environment variable for the service DockerID e.g. `$DOCKER_USER`
```

Comment:
IMO, We shouldn't be doing this manually. We should add them into CircleCI via Terraform, Store the secrets in Secrets Manager. Open to Ideas... any opinions @andrewpearce-digital ?

Reply:
There is also a Terraform Provider for Lastpass https://registry.terraform.io/providers/nrkno/lastpass/latest/docs

Reply:
Not used it, but should be simple enough to get going if we don't want to have the secret duplicated inside of Secrets Manager

Reply:
if there is a way to automate this, i'd prefer it in the long run.

Reply:
That is an excellent idea, though I'd consider creating the docker access tokens in terraform and adding them to the circle ci project the way we do with aws access keys.

Until we do that, this doc stands as the way it has been done

I may look at that approach when I pick up my ticket to do this for ual

Reply:
happy to look at this with you @andrewpearce-digital